### PR TITLE
[cmake] Internal depends download macro refactor

### PR DESCRIPTION
**Pull request by fuzzard**
## Description
Refactor some repetitive code into a macro.
Essentially any cmake internal dependency uses the same code to set download url.

## Motivation and context
Part of my work looking at moving tools/depends to a full cmake build. One avenue was to reuse the same methods used to do an internal dependency build. To simplify this i was looking at ways we can refactor out the current code to a generic macro/function.
For those eagle eyed viewers at home, you may notice the generalisation of the externalproject_add as much as i can. The next step would be to template that out into a macro/function.
This is part of that work. Whether i continue with tools/depends via this route is TBD, but at least minimising code duplication is good imo.

Of note, the libdvd and ffmpeg modules use different base_urls due to their hosting being on github and not our mirrors. Therefore they arent changed to use the macro. In the long run, i think we should move them to our mirrors to unify everything.

A couple of packages also had duplicate find_package_handle_standard_args calls. We can set all the variables we need in our build path, so that call has been pulled out where appropriate.

@howie-f if you have time to test, i know you've been using these internal depends from time to time

## How has this been tested?
osx various internal depends paths.

## What is the effect on users?
N/A

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by fuzzard** [comment]
Reviewers might want to use https://github.com/xbmc/xbmc/pull/20905/files?w=1 as a couple of modules have a bunch of whitespace changes to see the functional changes more easily

**Comment by howie-f** [comment]
it seems to have issues with internal `FMT`

```
CMake Error at /usr/share/cmake-3.18/Modules/FindPackageHandleStandardArgs.cmake:165 (message):
  Could NOT find Fmt (missing: FMT_VERSION)
Call Stack (most recent call first):
  /usr/share/cmake-3.18/Modules/FindPackageHandleStandardArgs.cmake:458 (_FPHSA_FAILURE_MESSAGE)
  cmake/modules/FindFmt.cmake:71 (find_package_handle_standard_args)
  cmake/scripts/common/Macros.cmake:372 (find_package)
  cmake/scripts/common/Macros.cmake:386 (find_package_with_ver)
  CMakeLists.txt:186 (core_require_dep)
```
but `CrossGUID`, `Dav1d`, `Udfread`, `FlatBuffers` look ok

**Comment by fuzzard** [comment]
Cheers @howie-f pushed the fix for the fmt version failure.

File: cmake/modules/FindCrossGUID.cmake [hunk 1]
```diff
@@ -2,30 +2,24 @@ if(ENABLE_INTERNAL_CROSSGUID)
...
-                      URL_HASH ${CROSSGUID_HASH}
+  set(CROSSGUID_VER ${${MODULE}_VER})
+
+  externalproject_add(${MODULE_LC}
+                      URL ${${MODULE}_URL}
+                      URL_HASH ${${MODULE}_HASH}
                       DOWNLOAD_DIR ${TARBALL_DIR}
-                      PREFIX ${CORE_BUILD_DIR}/crossguid
+                      DOWNLOAD_NAME ${${MODULE}_ARCHIVE}
+                      PREFIX ${CORE_BUILD_DIR}/${MODULE_LC}
```
Comment (wsnipex) [inline cmake/modules/FindCrossGUID.cmake:22]: all those indirect references like ${${MODULE}_XXX} are unnecessarily hard to read.
Why not use CROSSGUID_XXX instead?

File: cmake/scripts/common/ModuleHelpers.cmake [hunk 2]
```diff
@@ -48,3 +50,20 @@ function(get_archive_name module_name)
...
+macro(SETUP_BUILD_VARS)
+  get_archive_name(${MODULE_LC})
+  string(TOUPPER ${MODULE_LC} MODULE)
+
+  # allow user to override the download URL with a local tarball
+  # needed for offline build envs
+  if(${MODULE}_URL)
+    get_filename_component(${MODULE}_URL "${${MODULE}_URL}" ABSOLUTE)
+  else()
+    set(${MODULE}_URL ${${MODULE}_BASE_URL}/${ARCHIVE})
```
Comment (wsnipex) [inline cmake/scripts/common/ModuleHelpers.cmake:64]: I must be blind, can't find where this is set.. should that be ${${MODULE}_ARCHIVE}?

File: cmake/modules/FindCrossGUID.cmake (hunk 1 above)
Comment (fuzzard) [reply cmake/modules/FindCrossGUID.cmake:22]: The end goal is to template everything out. A single helper function/macro. So the less "hardcoded" names, the easier that is to do.

File: cmake/scripts/common/ModuleHelpers.cmake (hunk 2 above)
Comment (fuzzard) [reply cmake/scripts/common/ModuleHelpers.cmake]: Good spot. fixed

File: cmake/modules/FindCrossGUID.cmake (hunk 1 above)
Comment (fuzzard) [reply cmake/modules/FindCrossGUID.cmake:22]: An example of some of the testing/WIP stuff i did for other dependencies can be spotted here for an idea about the templating

https://github.com/fuzzard/xbmc/commits/cmake_python

Just keep in mind, its just a dump of stuff i was playing with. Theres a few depends (libass, python3 + modules, cdio, etc) that i used. Essentially the format i was trying was pull out the build stuff from the FindModules into their own folder, existing FindModules then call the buildmodule if desired, and the intent would be to just glob everything else in and work through their dependences. its EXTREMELY wip, and no where near that state, and im not sure i'll even keep going that path as yet, but was just something to play with.

**Review by howie-f — COMMENTED** [review]
one mismatch left as far as i can tell

https://github.com/xbmc/xbmc/blob/b70c1598bbd0ec13263694554d262d964c97bd8d/tools/depends/target/libudfread/LIBUDFREAD-VERSION#L1-L4

the current SHA512 for `libudfread-1.1.2.tar.gz` on mirrors is
`3069feb5db40288beb5b112b285186162a704f0fdd3cf67a17fd4eeea015f2cfcfbb455b7aa7c3d79d00fd095a3fd11cffc7b121dce94d99c3b06a509a8977d2`

--

hash of gtest is fine, but this configure message:
```
-- Configuring done
CMake Error at CMakeLists.txt:401 (add_dependencies):
  The dependency target "gtest" of target "kodi-test" does not exist.
```

**Comment by fuzzard** [comment]
Thanks again @howie-f 

Fixup for gtest is in, and the hash mismatch is due to a couple things being merged based on different things. Have a commit to update the hash from the 1.1.0 tarball to the current version.

**Review by howie-f — APPROVED** [review]
good to go from my pov. nice

File: tools/depends/target/libudfread/LIBUDFREAD-VERSION [hunk 3]
```diff
@@ -1,4 +1,4 @@
 LIBNAME=libudfread
 VERSION=1.1.2
 ARCHIVE=$(LIBNAME)-$(VERSION).tar.gz
-SHA512=340a03fe90d26a8a5c78e1e4f558a0b448a14332a661494f44af7de3e6c98cd219125e19f69d2a611ecb4870648a5d5b55d794e665eb8ec4192c0b499a0701ed
+SHA512=3069feb5db40288beb5b112b285186162a704f0fdd3cf67a17fd4eeea015f2cfcfbb455b7aa7c3d79d00fd095a3fd11cffc7b121dce94d99c3b06a509a8977d2
```
Comment (garbear) [inline tools/depends/target/libudfread/LIBUDFREAD-VERSION:4]: did the library change or was the hash just wrong?

Comment (fuzzard) [reply tools/depends/target/libudfread/LIBUDFREAD-VERSION:4]: Different PRS were merged based on different versions. was a bump, but the hash or was based off the old version

**Review by wsnipex — APPROVED** [review]

File: cmake/scripts/common/ModuleHelpers.cmake [hunk 4]
```diff
@@ -48,3 +50,20 @@ function(get_archive_name module_name)
...
+  # needed for offline build envs
+  if(${MODULE}_URL)
+    get_filename_component(${MODULE}_URL "${${MODULE}_URL}" ABSOLUTE)
+  else()
+    set(${MODULE}_URL ${${MODULE}_BASE_URL}/${${MODULE}_ARCHIVE})
+  endif()
+  if(VERBOSE)
+    message(STATUS "${MODULE}_URL: ${${MODULE}_URL}")
+  endif()
+endmacro()
```
Comment (howie-f) [inline cmake/scripts/common/ModuleHelpers.cmake:69]: a no blocking missing newline

Comment (fuzzard) [reply cmake/scripts/common/ModuleHelpers.cmake:69]: This is fixed in the follow up PR, so ill leave as is for now

**Pull request merged**